### ISSUE TITLE
Keyboard Shortcuts Dialog (Compact)

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#320]
The design conversation was started in https://github.com/phetsims/balloons-and-static-electricity/issues/263

Much discussion about general keyboard shortcut dialog design has occurred since. Dialogs for several other sims have taken a compact approach. Taking that into account I have drafted new content for a compact version for Balloons and Static Electricity. It lives in the same google doc as the more verbose version: [BASE: Keyboard Shortcuts](https://docs.google.com/document/d/1vk8_N3GuyC0o_Nl8HpBUu0tvVuWyWQ8FiksexPs_qBk/edit#)

I have two draft versions of the content to help get the visual design re-started.

@arouinfar  and @emily-phet, do have any thoughts or preferences on the following two drafts?

**Compact Version 1**
H1: Keyboard Shortcuts
H2: Grab or Release Balloon
- Grab or release balloon: [Space] or [Enter]

H2: Jump Grabbed Balloon
- Jump close to Sweater: [J]+[S]
- Jump to Wall: [J]+[W]
- Jump to Near wall: [J]+[N]
- Jump to Center: [J]+[C]

H2: Move Grabbed Balloon
- Move grabbed balloon: [Arrow keys] or [WASD]
- Move slower: [Shift] + [Arrow keys] or [Shift] + [WASD] (stacked)

H2: General Navigation
- Move to next item or group: [Tab] 
- Move to previous item or group: [Shift] + [Tab]
- Exits a dialog: [ESC]

**Compact Version 2**
H1: Keyboard Shortcuts
H2: Grab or Release Balloon
- Grab or release balloon: [Space] or [Enter]

H2: Move or Jump Grabbed Balloon
- Move grabbed balloon: [Arrow keys] or [WASD]
- Move slower: [Shift] + [Arrow keys] or [Shift] + [WASD] (stacked)
- Jump close to Sweater: [J]+[S]
- Jump to Wall: [J]+[W]
- Jump to Near wall: [J]+[N]
- Jump to Center: [J]+[C]

H2: General Navigation
- Move to next item or group: [Tab] 
- Move to previous item or group: [Shift] + [Tab]
- Exits a dialog: [ESC]

**Comment by terracoda** [comment]
@arouinfar I put the Move and Jump together, so they could be communicated as a list of 6 items, encouraging a non-visual user to read on.

**Comment by terracoda** [comment]
Oops forgot the the Arrow Keys for navigating group items. Here's Version 2 again:

H1: Keyboard Shortcuts
H2: Grab or Release Balloon
- Grab or release balloon: [Space] or [Enter]

H2: Move or Jump Grabbed Balloon
- Move grabbed balloon: [Arrow keys] or [WASD]
- Move slower: [Shift] + [Arrow keys] or [Shift] + [WASD] (stacked)
- Jump close to Sweater: [J]+[S]
- Jump to Wall: [J]+[W]Jump to Near wall: [J]+[N]
- Jump to Center: [J]+[C]

H2: General Navigation
- Move to next item or group: [Tab] 
- Move to previous item or group: [Shift] + [Tab]
- Move between items in a group: [Left Right Arrows] or [Up Down Arrows]
- Exits a dialog: [ESC]

**Comment by terracoda** [comment]
@arouinfar, please reassign me when you want me to have a look.

**Comment by arouinfar** [comment]
@terracoda the text looks great! I'll mock it up.

**Comment by arouinfar** [comment]
@terracoda I've mocked up two options for the dialog following the compact style. The first option groups all balloon controls together in one column. The second option has the balloon sections side-by-side and is a bit more compact. Visually, I prefer the second option, but the first option may be more advantageous for screen readers.

![screen shot 2017-11-01 at 3 41 41 pm](https://user-images.githubusercontent.com/8419308/32299434-3069b952-bf1b-11e7-825d-09a6cae144ac.png)
![screen shot 2017-11-01 at 3 40 41 pm](https://user-images.githubusercontent.com/8419308/32299381-0f23f8f2-bf1b-11e7-9759-d54fd0f0506c.png)

@terracoda can you review? I can't remember if we reached a consensus about whether or not to try out the alternative layouts from the design crit, so please let me know if you would like me to pursue other options.

**Comment by terracoda** [comment]
@arouinfar, this is an interesting layout problem. I like the look of the even columns, but the reading order is less obvious in that example. The reading order should go from Grab or Release to Move or Jump to General Navigation.

Caren Watkins provided an interesting approach during the design crit. Of course, the actual content was different then, but do you think you could try to visually connect the Grab or Release section with the Move  or Jump section like Caren has done below.

<img width="650" alt="base-kbdialog" src="https://user-images.githubusercontent.com/1902116/32304575-5b42291a-bf53-11e7-9250-7df0b6498b90.png">

**Comment by terracoda** [comment]
@arouinfar, it doesn't have to be exactly the same as Caren's example, but some kind of visual cueing  (lines, arrows or white space) that help to move the eye to the right, and then back to the bottom left for General Navigation.

**Comment by terracoda** [comment]
@arouinfar, if that doesn't work, we could bring back a heading for the Jump section, so there are 4 sections rather than 3.

**Comment by terracoda** [comment]
@arouinfar I think the jump key icons would look nicer at the front.

**Comment by terracoda** [comment]
H1: Keyboard Shortcuts
H2: Grab or Release Balloon
- Grab or release balloon: [Space] or [Enter]

H2: Move or Jump Grabbed Balloon
- Move grabbed balloon: [Arrow keys] or [WASD]
- Move slower: [Shift] + [Arrow keys] or [Shift] + [WASD] (stacked)
- [J]+[S] jumps close to sweater.
- [J]+[W] jumps to wall. 
- [J]+[N] jumps to near wall.
- [J]+[C] jumps to center. 

H2: General Navigation
- Move to next item or group: [Tab] 
- Move to previous item or group: [Shift] + [Tab]
- Move between items in a group: [Left Right Arrows] or [Up Down Arrows]
- Exits a dialog: [ESC]

**Comment by terracoda** [comment]
@arouinfar, what do you think of this layout that aligns jump keys to left and uses white space to de-emphasize Gen Nav section?
![keyboard-dialog](https://user-images.githubusercontent.com/1902116/32638253-068507ac-c599-11e7-9044-70fb7444f8aa.jpg)

**Comment by terracoda** [comment]
@arouinfar, obviously not the same font in my re-make ;-) and oops I see some copy paste mistakes - reposting.
![keyboard-dialog](https://user-images.githubusercontent.com/1902116/32638366-86836656-c599-11e7-9ca2-b676949db5a2.jpg)

**Comment by terracoda** [comment]
@jessegreenberg, here's what I think the Keyboard Shortcuts should be in the accessible PDOM Dialog. I re-arranged the words to follow the visual order.

I din't do it myself because some of the strings need to be removed, and m=some may need to be renamed.

```js
    // strings for keyboard shortcuts help content
    keyboardHelpDialogString: 'Keyboard Shortcuts',
    grabBalloonString: 'Grab or Release Balloon',
    grabDescriptionString: 'Grab or release balloon with Space or Enter keys.',
    hotkeysJumpingString: 'Move or Jump Grabbed Balloon',
// REMOVE next line
    draggingDescriptionString: 'Arrow keys drag and rub grabbed balloon.',
//New moving instructions
    draggingDescriptionWASDString: 'Move grabbed balloon up, left, down, or right with Arrow keys or with letter keys W, A, S, or D.',
//New move slower instructions
    addShiftString: 'Move slower with Shift plus Arrow keys or Shift plus letter keys W, A, S, or D.',
    JSString: 'J plus S jumps close to sweater.',
    JWString: 'J plus W jumps to wall.',
    JNString: 'J plus N jumps to near wall.',
    JCString: 'J plus C jumps to center.',
//REMOVE next line
    keysForDraggingAndRubbingString: 'Move Grabbed Balloon',
//Edited gen nav items to follow visual order
    generalNavigationString: 'General Navigation',
    tabString: 'Move to next item with Tab key.',
    shiftTabString: 'Move to previous item with Shift plus Tab key.', 
//New group navigation item.
   groupNavigationString: 'Move between items in a group with Left and Right Arrow keys or Up and Down Arrow keys.',
   escString: 'Exit a dialog with Escape key.',
```

**Comment by terracoda** [comment]
@jessegreenberg, do you think you could update the PDOM Dialog without also doing the visual dialog?

**Comment by terracoda** [comment]
@jessegreenberg, not the _Move or Jump Grabbed Balloon_ section would have a list with 6 items.

**Comment by jessegreenberg** [comment]
@terracoda done with the commits above, including a first pass at the visual content as well. It can be found at master, here is what it looks like now:
![capture](https://user-images.githubusercontent.com/6396244/32758000-7c0a152c-c89f-11e7-81ff-4dd1f02bf933.PNG)

And the accessible content:
![capture2](https://user-images.githubusercontent.com/6396244/32758003-7f2dc44c-c89f-11e7-8477-6512f77d8e11.PNG)

**Comment by terracoda** [comment]
@jessegreenberg, that's awesome!
I will check phettest tomorrow.

**Comment by jessegreenberg** [comment]
Removing my assignment for now.

**Comment by jessegreenberg** [comment]
@terracoda @arouinfar at keyboard navigation meeting on 11/14/17, there was some discussion about  changing the design of this dialog again. I recommend that this be done before we publish a keyboard navigable sim, including RIAW (https://github.com/phetsims/QA/issues/66). Once a sim is published, these strings will be translatable and they will become much more difficult to change. This does not include strings for the PDOM, those will not be translatable for a while.

**Comment by terracoda** [comment]
@jessegreenberg, excellent point.

**Comment by terracoda** [comment]
@jessegreenberg, re https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-344114563, the order of content is i n the accessible version is incorrect. The Move and Jump Balloon section needs to come before General Navigation section.

The white space in the visual design hints at a left-to-right-and-then-down exploration, and in the non-visual environment, the exploration is linear from top to bottom in order of importance.

**Comment by jessegreenberg** [comment]
Thanks @terracoda, general navigation section is now last.

**Comment by terracoda** [comment]
@arouinfar 
 drafting an alternative text version to one that is implemented above for your review. This version fronts the actions. I'm not sure it matters a tonne. I like to front the most important information. Both the key press and what it does are important pieces of information. I'm not sure one or the other is a clear winner.

The visual and non-visual representations do not have to have the same order. We should strive to provide the information in the order that is best for the groups of students accessing the content. 

Looking forward to your thoughts based on everything been done, and the bigger PhET picture.

I am going to tag @emily-phet, too, in case she has some thoughts.

**H1: Keyboard Shortcuts**
**H2: Grab or Release Balloon**
- Space or Enter key grab and release balloon. 

**H3: Move or Jump Grabbed Balloon**
- Arrow keys, or Letter keys, W, A, S, or D, move grabbed balloon up, left, down, or right. 
- Shift plus an Arrow key (up, left, down, or right) or Shift plus a letter key (W, A, S, or D) move slower.
- J plus S jumps close to sweater.
- J plus W jumps to wall.
- J plus N jumps to near wall.
- J plus C jumps to center.

**H2: General Navigation**
- Tab key moves to next item.
- Shift plus Tab moves to previous item. 
- Arrow keys, left and right, or up and down, move between items in a group.
- Escape key exits a dialog.

**Comment by terracoda** [comment]
@jessegreenberg, I just noticed a small mistake in the text version. Sorry, I missed it.
- J plus W jumps close to wall.

Should be:
- J plus W jumps to wall.

**Comment by jessegreenberg** [comment]
Thanks @terracoda,  fixed in https://github.com/phetsims/balloons-and-static-electricity/commit/09aaf4f4d2e8b1b7d41734e6e3f771b148222e06

**Comment by terracoda** [comment]
@arouinfar, there is also a draft for Ohm Law in https://github.com/phetsims/ohms-law/issues/100

And that text also follows the visual design putting the  explanation of the key press before the key press.

Changes now will affect BASE, JT, OL, and RIAW.

If you are good with version one that has already been quickly implemented by JG. I am good with that approach until we know more what users want as text versions.

**Comment by terracoda** [comment]
@arouinfar, do you remember when we decided to go with putting the icons at the end? I was just looking at BASE's original version over in #263

**Comment by arouinfar** [comment]
@terracoda can you say a bit about your motivation for reversing the order on the jump keys?

I think we should have a consistent ordering in the dialog, with the key icons always coming before or after the text. For short commands, like the jump keys, fore-fronting the key icon looks clean. However, I think most commands are harder to read and look cluttered with key icon on the left, particularly the first two "move" commands which will appear in most dialogs.

I think fore-fronting the text has a few advantages
* The dialog is easier to skim with the text left-aligned. Users can quickly scan the dialog for a command they are looking for, then look at the corresponding key icon(s).
* Screen reader users will hear the content ordered as [Command] [Key(s)], so it's nice (though I realize not necessary) to keep the same visual ordering.
* Overall, the content looks cleaner (IMO)

 
<img width="798" alt="screen shot 2017-11-17 at 9 51 22 am" src="https://user-images.githubusercontent.com/8419308/32958848-ec290126-cb7c-11e7-99f7-9bf0eda98f72.png">
<img width="800" alt="screen shot 2017-11-17 at 9 51 27 am" src="https://user-images.githubusercontent.com/8419308/32958849-ec578744-cb7c-11e7-8289-d7800c64db3b.png">

**Comment by terracoda** [comment]
@arouinfar, I was just going to ping you about moving the icons to the left side.

**Comment by terracoda** [comment]
Starting a consolidated issue on the a11y-research repo.

**Comment by emily-phet** [comment]
@terracoda @arouinfar 
- I just asked around to some folks who hadn't seen this before. It does seem that the text left/image right seems to be more understandable. @arouinfar can you make a quick version where the text is to the left, and the images are on the right, but the images are right aligned? 

One concern I have is that when the text is translated, we could run into layout issues with having images placed based on the location of the last word in the string. If it worked to have the images right aligned and the text left aligned, I think that might end up being better.

**Comment by terracoda** [comment]
@emily-phet, @arouinfar mocked things up over in https://github.com/phetsims/a11y-research/issues/62.

Can you comment there?

**Comment by jessegreenberg** [comment]
Since we have come to decisions about the dialog layout in  phetsims/a11y-research#62, I think we are ready to  continue here. @terracoda @arouinfar can you please update this issue with the new look for the dialog after working through the general issue?

**Comment by arouinfar** [comment]
Here's a mockup of the BASE dialog following the style decided upon in https://github.com/phetsims/a11y-research/issues/62. I've kept the General Navigation section aligned with the jump keys as @terracoda suggested earlier. 

![image](https://user-images.githubusercontent.com/8419308/33100939-79a2f8dc-ced3-11e7-846f-9d4d88d3f03e.png)

**Comment by jessegreenberg** [comment]
Looks great @arouinfar, thanks. It looks like we will need a way to align keys of different groups of help content, since "Space or Enter" icon needs to aling with "Tab" icon.

**Comment by terracoda** [comment]
Looks great @arouinfar, but with the new column layout for the controls, it might look better with all the Grab and Move content on the left side. Here's a whipped up example. Could you move the Move section back under the Grab and Release section. That should prevent the Space and Enter keys from looking like they are drowning in a sea of blue.

Add sufficient heading spacing, my mock-up is very rough.

![base-kb-dialog](https://user-images.githubusercontent.com/1902116/33144236-e4a88e94-cf96-11e7-9c22-00bcdeab8356.jpg)
.

**Comment by terracoda** [comment]
@arouinfar assigning back to you for one more go. Hopefully, the last one. Do you thinking moving the Move and Jump stuff back over to the left side works better?

**Comment by arouinfar** [comment]
@terracoda, I think it flows a bit better if the Grab and Move sections are in the same column, even if the dialog does end up a bit larger. I've added a bit more negative space between the Grab and Move sections, but otherwise, your mockup was pretty spot-on. :)

<img width="861" alt="screen shot 2017-11-22 at 3 56 55 pm" src="https://user-images.githubusercontent.com/8419308/33153253-3ae40df4-cf9e-11e7-9bc9-884a66ef616a.png">


Off to @jessegreenberg for implementation/reassignment.

**Comment by terracoda** [comment]
Awesome @arouinfar!

**Comment by terracoda** [comment]
Oh @jessegreenberg , one text issue with the visual mock-up.
- "Jump close to wall", should be "Jump to wall"

**Comment by arouinfar** [comment]
Good catch @terracoda. Thanks!

**Comment by arouinfar** [comment]
I updated the mockup in https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-346495639 to fix the typo.

**Comment by terracoda** [comment]
@arouinfar, you are quick as lightening!

And for the non-visual environment, I need to switch around the Jump phrases, so they read like this:
- Jump close to sweater with J plus S key.
- Jump to wall with J plus S key.
- Jump to near wall with J plus N key.
- Jump to center with J plus N key.

I think all the other text in comment https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-344114563 is good.

**Comment by terracoda** [comment]
@jessegreenberg, ready for implementation, and unassigning myself.

**Comment by terracoda** [comment]
@jessegreenberg, do you think you could implement the accessible content for the keyboard shortcuts dialog next. I will need it to schedule interviews for this sim.

**Comment by jessegreenberg** [comment]
@terracoda help content has been updated, here it is in the a11y view:
![capture](https://user-images.githubusercontent.com/6396244/36622980-d90af610-18bd-11e8-97d9-7cc9d1d93b6e.PNG)

Dialog also now looks like the mockup in https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-346495639.  Here is a dev version, @arouinfar @terracoda could you please review?

Dev version:  https://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.4.0-dev.12/phet/balloons-and-static-electricity_en_phet.html

**Comment by arouinfar** [comment]
Looks great @jessegreenberg!

**Comment by jessegreenberg** [comment]
Thanks @arouinfar and @terracoda. I recall a previous discussion where we decided to have the  close button last. I think ideally, the AT should read the dialog content in full  when it is open, so maybe we should also figure out why it isn't doing that?

**Comment by jessegreenberg** [comment]
Here is the reason why the close button is last: https://github.com/phetsims/john-travoltage/issues/219

**Comment by terracoda** [comment]
@jessegreenberg, I remember that conversation as well, and remember advising that the close button should go last.

**Comment by terracoda** [comment]
@jessegreenberg, in my reading of phetsims/john-travoltage#219, I think the assumption was that there focus want placed on some focusable content before the close button.

Is that your reading as well?

**Comment by jessegreenberg** [comment]
@terracoda I think the assumption was focus should go to the first focusable item in the dialog, and the close button should be last. For dialogs where the close button is the only focusable element, focus would go there.

I see there is a bug in KeyboardHelpContent preventing aria-describedby from being set on the Dialog content as it should be.

When this is fixed NVDA reads all content in the dialog from top to bottom, then reads "Close, button". VoiceOver reads the Dialog title, then the close button, then each heading in the Dialog. @terracoda does this seem acceptable?

**Comment by terracoda** [comment]
@jessegreenberg, thanks for checking. I think it does make sense for the close button to come last as long as the user hears the content at least once. If the hear the content and then end on the close button, it seems to me that it would be more intuitive to use the left arrow key to go back and re-read.

We should watch this carefully in interviews to find out how users are accessing the content in the dialog.

**Comment by jessegreenberg** [comment]
Sounds good @terracoda, fixed in the above commit. If you decide it is best to have the close button first, just let me know. @terracoda anything else for this issue?

**Comment by terracoda** [comment]
@jessegreenberg, I will review this change on PhET Test. Thanks and report back asap.

**Comment by terracoda** [comment]
@jessegreenberg, I think we might need to move the close button to the top of the dialog. 
When I start reading with the arrow keys, my focus goes from close button to end of list.

In an interview, today, the user had trouble with the latest dev version (13, I think) and they closed the dialog only after reading about General Navigation only.

I am going to flip my original guidance for this dialog. 
Since the close button is the only focusable item in the dialog, it is easily found with the Tab key no matter where it is. However, if it is at the bottom, reading with the Arrow keys seems unintuitive as the cursor is already at the bottom.

If the close button, is first (just before or just after the H1) then pressing the cursor keys to read would start at the top.

Any thoughts?
It's not working for all users as-is in dev 13 :-)

**Comment by jessegreenberg** [comment]
Sounds good @terracoda, Ill move the close button to  the top of the dialog so it is at the top of other content for the virtual cursor.

**Comment by jessegreenberg** [comment]
@terracoda change made in the above commit. By the way, it sounds really great with NVDA + Firefox. NVDA puts focus on the close button, but reads the rest of the Dialog content from top to bottom. At the end, the AT reads the "Close button" but the rest of the dialog content can be read from top to bottom. This was a good idea @terracoda. Closing.

**Comment by jessegreenberg** [comment]
Reopening, when I changed the layout for https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-346495639 @phet-steele noticed that I didn't remove the vertical space between the sectons on the left side of the dialog.

<img width="716" alt="screen shot 2018-04-06 at 10 44 06 am" src="https://user-images.githubusercontent.com/6396244/38434395-643a527e-398b-11e8-9103-e092210fceb0.png">

**Comment by terracoda** [comment]
Her's the comment with the final visual layout in case you need it. This is a long issue.
https://github.com/phetsims/balloons-and-static-electricity/issues/320#issuecomment-346495639

**Comment by terracoda** [comment]
Please re-assign to me if there are questions, thanks.

**Comment by jessegreenberg** [comment]
I reduced  the spacing a bit. @terracoda @phet-steele here is the look  now:
![capture](https://user-images.githubusercontent.com/6396244/38580764-12a533b2-3cc8-11e8-821a-11400125bbff.PNG)

If this seems reasonable please close this issue. Otherwise reassign.

**Comment by phet-steele** [comment]
I like it.

**Comment by terracoda** [comment]
I like the visual, too.

However, when I open the dialog with VoiceOver, I get some unexpected verbosity (see screen shot). It is normal for a screen reader to read out the number of headings or items, but the case below, it says 3 items (which is incorrect) and then it reads out a mix of content and actual heading content.

@jessegreenberg , what do you get with NVDA? 

I don't remember VoiceOver being this verbose.

<img width="915" alt="screen shot 2018-04-12 at 21 22 01" src="https://user-images.githubusercontent.com/1902116/38709925-406b0c14-3e98-11e8-9486-07ff93acf323.png">

This visual design looks great, so maybe we should close this issue and start a new one?

**Comment by terracoda** [comment]
Since I am the only one assigned, I'll go ahead and start a new issue about the strange VO verbosity.
